subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt_hash_checksum_sha256.c

```diff
@@ -108,30 +64,41 @@ static int fs_mgmt_hash_checksum_sha256(struct fs_file_t *file, uint8_t *output,
 		bytes_read = fs_read(file, buffer, read_size);
 
 		if (bytes_read < 0) {
-			/* Failed to read file data, pass generic unknown error back */
-			rc = MGMT_ERR_EUNKNOWN;
-			goto error;
+			/* Failed to read file data */
+			goto teardown;
 		} else if (bytes_read > 0) {
-			if (mbedtls_md_update(&mbed_hash_ctx, buffer, bytes_read) != 0) {
-				rc = MGMT_ERR_EUNKNOWN;
-				goto error;
+#if PSA_IMPLEMENTATION
+			op_ret = psa_hash_update(&psa_hash_ctx, buffer, bytes_read);
+#else
+			op_ret = mbedtls_sha256_update(&mbed_hash_ctx, buffer, bytes_read);
+#endif
+			if (op_ret != SUCCESS_VALUE) {
+				goto teardown;
 			}
 
 			*out_len += bytes_read;
 		}
 	} while (bytes_read > 0 && *out_len < len);
 
 	/* Finalise SHA256 hash calculation and store output in provided output buffer */
-	if (mbedtls_md_finish(&mbed_hash_ctx, output) != 0) {
-		rc = MGMT_ERR_EUNKNOWN;
+#if PSA_IMPLEMENTATION
+	op_ret = psa_hash_finish(&psa_hash_ctx, output, SHA256_DIGEST_SIZE, &read_size);
+#else
+	op_ret = mbedtls_sha256_finish(&mbed_hash_ctx, output);
+#endif
+	if (op_ret == SUCCESS_VALUE) {
+		rc = 0;
 	}
 
-error:
-	mbedtls_md_free(&mbed_hash_ctx);
+teardown:
+#if PSA_IMPLEMENTATION
+	psa_hash_abort(&psa_hash_ctx);
+#else
+	mbedtls_sha256_free(&mbed_hash_ctx);
+#endif
```

Comment:
looking all these ifdefs, wouldn't be cleaner having two distinct implementations ? like currently exists for TinyCrypt and mbedTLS ?

Reply:
The implementations have so much in common that I thought it would be a bit of a pity to duplicate the code. (And the `#ifdef`s are partly the reason why I introduced the `PSA_IMPLEMENTATION` macro, to make them more readable.)

What do the others think?

Reply:
I think that, by the signatures of the functions, we could just define "API" here in form:
```
typedef  ...  hash_ctx // this would define psa_hash_operation_t or mbedtls_sha256_context type object
int hash_setup(hash_ctx *ctx) // optionally taking hash type, but I would not do that, this is local thing.
int hash_update(hash_ctx *ctx, uint8_t *buff, size_t buf_size)
int hash_finish(hash_ctx *ctx, uint8_t *buff)
int hash_cleanup(hash_ctx *ctx) // serves as abort/free
```

Group above in `#ifdef CONFIG_MBEDTLS_PSA_CRYPTO_CLIEN, #else` block above or below `fs_mgmt_hash_checksum_sha256` and cut-down `#ifdefs` in that function. 

IMHO.